foreman/data_refinery_foreman/surveyor/sra.py

```diff
@@ -56,275 +54,137 @@ def source_type(self):
-        study_xml = ET.fromstring(response.text)
-
-        study = study_xml[0]
-        for child in study:
-            if child.tag == "DESCRIPTOR":
-                for grandchild in child:
-                    # STUDY_TYPE is the only tag which uses attributes
-                    # instead of the text for whatever reason
-                    if grandchild.tag == "STUDY_TYPE":
-                        metadata[grandchild.tag.lower()] = grandchild.attrib["existing_study_type"]
-                    else:
-                        metadata[grandchild.tag.lower()] = grandchild.text
-            elif child.tag == "STUDY_ATTRIBUTES":
-                for grandchild in child:
-                    key, value = SraSurveyor.parse_attribute(grandchild, "study_")
-                    metadata[key] = value
-            elif child.tag == "STUDY_LINKS":
-                for grandchild in child:
-                    for ggc in grandchild:
-                        if ggc.getchildren()[0].text == "pubmed":
-                            metadata["pubmed_id"] = ggc.getchildren()[1].text
-                            break
-            elif child.tag == "IDENTIFIERS":
-                for grandchild in child:
-                    if (
-                        # Check for GEO accessions. These live inside an
-                        # EXTERNAL_ID tag with namespace GEO
-                        grandchild.tag == "EXTERNAL_ID"
-                        and grandchild.attrib.get("namespace", "") == "GEO"
-                        and re.match(r"^GSE\d{2,6}", grandchild.text)
-                    ) or (
-                        # Check for ArrayExpress accessions. These live inside a
-                        # SUBMITTER_ID tag, but the namespace is not standardized
-                        grandchild.tag == "SUBMITTER_ID"
-                        and re.match(r"^E-[A-Z]{4}-\d{2,6}", grandchild.text)
-                    ):
-                        metadata["external_id"] = grandchild.text
-                        break
+    def gather_all_metadata(accession: str):
+        # check if experiment, if not fetch the sample, biosample and experiment
+        samples_metadata = SraSurveyor.gather_experiment_runs_metadata(accession)
 
-    @staticmethod
-    def gather_all_metadata(run_accession):
-        metadata = SraSurveyor.gather_run_metadata(run_accession)
+        if not re.match(r"^(S|E|D)RP\d+", accession):
+            experiment_accession = samples_metadata[0]["secondary_study_accession"]
+        else:
+            experiment_accession = accession
+
+        experiment_metadata = SraSurveyor.gather_experiment_metadata(experiment_accession)
+
+        if "accession" in experiment_metadata:
+            pubmed_metadata = SraSurveyor.gather_experiment_pubmed_metadata(
+                experiment_metadata["accession"]
+            )
+            if len(pubmed_metadata) > 0:
+                experiment_metadata["pubmed_id"] = pubmed_metadata["Source Primary Accession"]
```

Comment:
This is a limitation of our current db structure, which I would like to avoid changing at this time. I do think you are right that we would want to handle more than one.